fdbserver/Restore.actor.cpp

```diff
@@ -18,86 +18,606 @@
+	//NUM_APPLIERS = 1;
+	NUM_APPLIERS = g_network->isSimulated() ? 3 : NUM_APPLIERS;
+	MIN_NUM_WORKERS  = NUM_LOADERS + NUM_APPLIERS;
+	FastRestore_Failure_Timeout = 3600; // seconds
+	loadBatchSizeMB = g_network->isSimulated() ? 1 : loadBatchSizeMB; // MB
+	loadBatchSizeThresholdB = loadBatchSizeMB * 1024 * 1024;
+	mutationVectorThreshold = g_network->isSimulated() ? 100 : mutationVectorThreshold; // Bytes // correctness passed when the value is 1
+	transactionBatchSizeThreshold = g_network->isSimulated() ? 512 : transactionBatchSizeThreshold; // Byte
+
+	// Debug
+	//loadBatchSizeThresholdB = 1;
+	//transactionBatchSizeThreshold = 1;
+
+	printf("Init RestoreWorkerConfig. min_num_workers:%d ratio_loader_to_applier:%d loadBatchSizeMB:%.2f loadBatchSizeThresholdB:%.2f transactionBatchSizeThreshold:%.2f\n",
+			MIN_NUM_WORKERS, ratio_loader_to_applier, loadBatchSizeMB, loadBatchSizeThresholdB, transactionBatchSizeThreshold);
+}
+
+
+// Restore Worker
+ACTOR Future<Void> commitRestoreRoleInterfaces(Reference<RestoreWorkerData> self, Database cx) {
+	state ReadYourWritesTransaction tr(cx);
+	// For now, we assume only one role per restore worker
+	ASSERT( !(self->loaderInterf.present() && self->applierInterf.present()) );
+
+	loop {
+		try {
+			tr.reset();
+			tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+			tr.setOption(FDBTransactionOptions::LOCK_AWARE);
+			ASSERT( !(self->loaderInterf.present() && self->applierInterf.present()) ); 
+			if ( self->loaderInterf.present() ) {
+				tr.set( restoreLoaderKeyFor(self->loaderInterf.get().id()), restoreLoaderInterfaceValue(self->loaderInterf.get()) );
+			}
+			if ( self->applierInterf.present() ) {
+				tr.set( restoreApplierKeyFor(self->applierInterf.get().id()), restoreApplierInterfaceValue(self->applierInterf.get()) );
 			}
+			wait (tr.commit() );
+			break;
+		} catch( Error &e ) {
+			printf("[WARNING]%s: commitRestoreRoleInterfaces transaction error:%s\n", self->describeNode().c_str(), e.what());
+			wait( tr.onError(e) );
 		}
 	}
 
-	//we are the leader
-	wait( delay(5.0) );
+	return Void();
+} 
+
+// Restore Worker
+ACTOR Future<Void> handleRecruitRoleRequest(RestoreRecruitRoleRequest req, Reference<RestoreWorkerData> self, ActorCollection *actors, Database cx) {
+	printf("[INFO][Worker] Node:%s get role %s\n", self->describeNode().c_str(),
+			getRoleStr(req.role).c_str());
+
+	while (self->isInProgress(RestoreCommandEnum::Recruit_Role_On_Worker)) {
+		printf("[DEBUG] NODE:%s handleRecruitRoleRequest wait for 1s\n",  self->describeNode().c_str());
+		wait(delay(1.0));
+	}
+	if ( self->isCmdProcessed(req.cmdID) ) {
+		req.reply.send(RestoreCommonReply(self->id(), req.cmdID));
+		return Void();
+	}
+	self->setInProgressFlag(RestoreCommandEnum::Recruit_Role_On_Worker);
+
+	if (req.role == RestoreRole::Loader) {
+		ASSERT( !self->loaderInterf.present() );
+		self->loaderInterf = RestoreLoaderInterface();
+		self->loaderInterf.get().initEndpoints();
```

Comment:
About `addRole`, did you mean the way  the existing FDB code adds the FDB roles (e.g., tLog) by using `startRole` function? 

I agree adding it will help debug. 
I'm a bit reluctant to add it in this PR because I want to have as little impact as possible to the existing FDB system as possible. But I'm happy to add it in the next PR>